### PR TITLE
Store profile picture in binary format

### PR DESCRIPTION
**Pull request by adolorosa**
I modified my files so that the profile picture can be stored as bytes in the db, but it doesn't seem to work - I get the following error: "TypeError: string argument without an encoding" and I'm not sure what to make of it.

### CODE REVIEW

**Comment by daniel-j-h** [comment]
Hm, do you have a more complete backtrace including line numbers?

My hunch is on how you convert the image to bytes. Can you try the following:
- use a BytesIO object which mimics a file but is memory-based
- save the image with its format (e.g. jpeg) into the memory-based BytesIO
- get the raw BytesIO content which is the (e.g. jpeg) encoded images

something similar to

```python3
import io

buf = io.BytesIO()
image.save(buf, format=image.format)
bts = buf.getvalue()
```

loading should then work similarly.

See
- https://docs.python.org/3/library/io.html#io.BytesIO
- https://docs.python.org/3/library/io.html#io.BytesIO.getvalue
- https://docs.python.org/3/library/stdtypes.html#bytes

**Comment by adolorosa** [comment]
Okay, so I used BytesIO and it seems to work as long as we're talking about converting the pic to bytes and storing it as a blob type in the db. But now I'm not sure how to display it in my template. The following code in my account route:
```
buf = io.BytesIO(current_user.image_file)
profile_pic = Image.open(buf)
```
seems to successfuly convert the bytes back to an image format, but the image doesn't get displayed in the browser. However, if I call
```
profile_pic.show()
```
it displays the correct image in my computer's default app, so I know the image got converted back and looks good. I just don't know how to make it appear in my template.

When I inspect the HTML code in my browser, the non-displaying image looks like this:
```
<img class="rounded-circle account-img mr-4" src="<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=117x125 at 0x7FC0BC162FD0>">
```

**Comment by daniel-j-h** [comment]
What you are seeing is the PIL Image object getting converted to its string representation, as in

    img = Image.open("myimg.jpg")
    print(img)

Instead you have to transfer the image data to the browser. I'd be best if you create a new endpoint, e.g.

    @app.route("/picture/<userid:int>")
    def picture(userid):
        pass  # implement as described below

where you then fetch the binary blob for the given userid from the database and then use Flask's `send_file` function to send the image with the correct media headers set.

Then in your template you can have an img html tag and in the src you can do something like

    

Then when the user's browser comes across this image tag it will fetch the image from your profile endpoint where you grab the blob from the database and send it back.

Check out
- http://flask.pocoo.org/docs/0.12/api/#flask.send_file
- https://en.wikipedia.org/wiki/Media_type

And an example how I'm doing this in a small Flask app: https://github.com/mapbox/robosat/blob/1d0cf506cde4600ab7063c238f7f9e25d65ba611/robosat/tools/serve.py#L128-L132

**Comment by adolorosa** [comment]
Thanks, I did it your way and it works now. Please let me know if it looks fine to you!

File: boredgamers/routes.py [hunk 1]
```diff
@@ -57,24 +59,31 @@ def logout():
...
+    i.save(buf, format=i.format)
+    bts = buf.getvalue()
+    return bts
+    
+
+@app.route("/picture/<int:userid>")
+def picture(userid):
+    user = User.query.filter_by(id=userid).first()
+    picture = user.image_file
+    return send_file(io.BytesIO(picture), mimetype="image/jpeg")
```
Comment (daniel-j-h) [inline boredgamers/routes.py]: I think above we allow users to upload jpgs or pngs. What will happen here if users upload a png file but we set the mime type always to jpg here? Should we either always convert to jpg (`PIL.Image.save` can do this for us) or set the mimetype based on loading the bytes back into a `PIL.Image` here and asking for its `.format`?

Comment (adolorosa) [reply boredgamers/routes.py]: Right, I missed that!
> set the mimetype based on loading the bytes back into a `PIL.Image` here and asking for its `.format`?

I'm not sure how to do that?

Comment (daniel-j-h) [reply boredgamers/routes.py]: You should be able to load the bytes into an `PIL.Image` with

    from PIL import Image
    image = Image.open(bytes)

and then you can ask for `image.format`.

**Comment by adolorosa** [comment]
I guess I'm doing it wrong, because it doesn't work. My code:
```
@app.route("/picture/<int:userid>")
def picture(userid):
    user = User.query.filter_by(id=userid).first()
    picture = user.image_file
    image = Image.open(picture)
    return send_file(io.BytesIO(picture), mimetype=image.format)
```
It results in the following error:

error

Wrapping it in BytesIO didn't help.

**Comment by daniel-j-h** [comment]
Hm

    Image.open(io.BytesIO(picture))

_should_ work. Can you have a look what the type is you get back from `user.image_file`, make sure it's actually bytes, and maybe print `type(picture)` and debug this a bit.

I think the mimetype can also not just be `image.format`; it has to be something like `image/jpeg` (and similar for png). See the mimetype wiki page and explanations about this.

**Comment by adolorosa** [comment]
I used this:

`Image.open(io.BytesIO(picture))`

and it works just fine :) I tested it with different image extensions and it seems to be ok. Thank you!

**Comment by daniel-j-h** [comment]
Hmm, you still have to send the correct mimetype, no? Right now you always send `image/jpeg` even for a png encoded image. Not sure why this works, maybe the browser auto-detects that the mimetype is wrong but still can decode the png image and then shows that? Anyway, you should probably fix it :D

**Comment by adolorosa** [comment]
Is this code all right then?
```
def picture(userid):
    user = User.query.filter_by(id=userid).first()
    picture = user.image_file
    image = Image.open(io.BytesIO(picture))
    return send_file(io.BytesIO(picture), mimetype=f"image/{image.format}")
```

**Comment by daniel-j-h** [comment]
Looks okay at least for jpeg and png - do you get jpeg or jpg back from `image.format`? You should probably just hardcode a lookup table from extension to mimetype, as in

```python
mimetypes = {
  "jpg": "image/jpeg",
  "jpeg": "image/jpeg",
   ..}
```

for jpegs, png, and whatever else we allow the user to upload.

**Pull request merged**